### PR TITLE
planner/cascades: add transformation rule EliminateOuterJoinBelowProjection

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve? 
This PR adds Transformation rule EliminateOuterJoinBelowProjection part of Outer join elimination in cascades planner(#13709).

### What is changed and how it works?
add a transformation rule EliminateOuterJoinBelowProjection.This rule tries to eliminate the outer join which below projection.

### Check List

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -1814,6 +1804,84 @@ func (r *EliminateOuterJoinBelowAggregation) isAggFuncAllDupAgnostic(agg *planne
...
+		return nil, false, false, nil
+	}
+
+	// only when proj only use the columns from outer table can eliminate outer join.
+	if !plannercore.IsColsAllFromOuterTable(proj.GetUsedCols(), outerUniqueIDs) {
+		return nil, false, false, nil
+	}
+
+	innerJoinKeys := join.ExtractJoinKeys(innerChildIdx)
+	contain, err := isInnerJoinKeysContainUniqueKey(innerGroup, innerJoinKeys)
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: I think `isInnerJoinKeysContainUniqueKey` and `prepareForEliminateOuterJoin` will only be used in `EliminateOuterJoinBelowProjection` and `EliminateOuterJoinBelowAggregation`. How about move these two methods to a new struct like:
```golang
type outerJoinEliminator struct {
}

func (e *outerJoinEliminator) prepareForEliminateOuterJoin(...){}

type EliminateOuterJoinBelowProjection struct {
  baseRule
  outerJoinEliminator
}
```

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: done

**Comment by SeaRise** [comment]
@francis0407 @lzmhhh123 PTAL

**Pull request merged**